src/Roave/SecurityAdvisories/AdvisorySources/GetAdvisoriesFromGithubApi.php

```diff
@@ -0,0 +1,135 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license.
+ */
+
+declare(strict_types=1);
+
+namespace Roave\SecurityAdvisories\AdvisorySources;
+
+use Generator;
+use Nyholm\Psr7\Request;
+use Psr\Http\Client\ClientExceptionInterface;
+use Psr\Http\Client\ClientInterface;
+use Psr\Http\Message\RequestInterface;
+use Roave\SecurityAdvisories\Advisory;
+use Safe\Exceptions\JsonException;
+use Safe\Exceptions\StringsException;
+use stdClass;
+use Webmozart\Assert\Assert;
+use function array_map;
+use function array_merge;
+use function end;
+use function explode;
+use function Safe\json_decode;
+use function Safe\json_encode;
+use function Safe\sprintf;
+
+final class GetAdvisoriesFromGithubApi implements GetAdvisories
+{
+    private const GRAPHQL_QUERY = 'query {
+            securityVulnerabilities(ecosystem: COMPOSER, first: 100 %s) {
+                edges {
+                    cursor
+                    node {
+                        vulnerableVersionRange
+                        package {
+                            name
+                        }
+                    }
+                }
+                pageInfo {
+                    hasNextPage
+                }
+            }
+        }';
+
+    /** @var ClientInterface */
+    private $client;
+
+    /** @var string */
+    private $token;
+
+    public function __construct(
+        ClientInterface $client,
+        string $token
+    ) {
+        $this->client = $client;
+        $this->token  = $token;
+    }
+
+    public function __invoke() : Generator
+    {
+        return yield from array_map(
+            static function (stdClass $item) {
+                return Advisory::fromArrayData(
+                    [
+                        'reference' => $item->node->package->name,
+                        'branches' => [['versions' => explode(',', $item->node->vulnerableVersionRange)]],
+                    ]
+                );
+            },
+            $this->getAdvisories()
+        );
+    }
+
+    /**
+     * @return stdClass[]
+     *
+     * @throws ClientExceptionInterface
+     * @throws JsonException
+     * @throws StringsException
+     */
+    private function getAdvisories() : array
+    {
+        $advisories = [];
+        $cursor     = null;
+        do {
+            $response        = $this->client->sendRequest($this->getRequest($cursor));
+            $data            = json_decode($response->getBody()->getContents());
+            $vulnerabilities = $data->data->securityVulnerabilities;
+
+            $advisories = array_merge($advisories, $vulnerabilities->edges);
+            if (! $hasNextPage = $vulnerabilities->pageInfo->hasNextPage) {
+                continue;
+            }
+
+            $cursor = end($advisories)->cursor;
```

Comment:
This cursor traversal is very confusing: what is being done here? Can it be rewritten without relying on it?